pkg/transport/manager.go

```diff
@@ -216,45 +216,43 @@ func (tm *Manager) SaveTransport(ctx context.Context, remote cipher.PubKey, tpTy
 		return nil, io.ErrClosedPipe
 	}
 
-	const tries = 2
-
-	var err error
-	for i := 0; i < tries; i++ {
-		mTp, err := tm.saveTransport(remote, tpType)
-		if err != nil {
-			return nil, err
-		}
-
-		if err = mTp.Dial(ctx); err != nil {
-			// TODO(nkryuchkov): Check for an error that underlying connection is not established
-			// and try again in this case. Otherwise, return the error.
-			pkTableErr := fmt.Sprintf("pk table: entry of %s does not exist", remote.String())
-
-			if err.Error() == pkTableErr {
-				mTp.wg.Wait()
-				delete(tm.tps, mTp.Entry.ID)
+TrySaveTp:
+	mTp, err := tm.saveTransport(remote, tpType)
+	if err != nil {
+		return nil, err
+	}
 
-				return nil, err
+	if err = mTp.Dial(ctx); err != nil {
+		// This occurs when an old tp is returned by 'tm.saveTransport', meaning a tp of the same transport ID was just
+		// deleted (and has not yet fully closed). Hence, we should close and delete the old tp and try again.
+		if err == ErrNotServing {
+			if closeErr := mTp.Close(); closeErr != nil {
+				tm.Logger.WithError(err).Warn("Closing mTp returns non-nil error.")
 			}
+			delete(tm.tps, mTp.Entry.ID)
+			goto TrySaveTp
```

Comment:
Could you please move to the for-loop instead of `goto`? All my nature just stands against it, even if it's mostly the same as the loop in such case